server/src/semanticTokensProvider.ts

```diff
@@ -0,0 +1,142 @@
+/* --------------------------------------------------------------------------------------------
+ * Copyright (c) 2023 Savoir-faire Linux. All rights reserved.
+ * Licensed under the MIT License. See License.txt in the project root for license information.
+ * ------------------------------------------------------------------------------------------ */
+
+import { type SemanticTokens, SemanticTokensBuilder, type SemanticTokensLegend } from 'vscode-languageserver/node'
+import { logger } from './lib/src/utils/OutputLogger'
+import { analyzer } from './tree-sitter/analyzer'
+import * as TreeSitterUtils from './tree-sitter/utils'
+import { type SyntaxNode } from 'web-tree-sitter'
+
+interface ParsedToken {
+  line: number
+  startCharacter: number
+  length: number
+  tokenType: string
+  tokenModifiers: string[]
+}
+
+const tokenTypes = new Map<string, number>()
+const tokenModifiers = new Map<string, number>()
+
+// https://code.visualstudio.com/api/language-extensions/semantic-highlight-guide
+export const TOKEN_LEGEND = {
+  types: {
+    function: 'function',
+    variable: 'variable',
+    parameter: 'parameter',
+    class: 'class',
+    number: 'number',
+    operator: 'operator',
+    string: 'string',
+    keyword: 'keyword'
+  },
+  modifiers: {
+    declaration: 'declaration',
+    deprecated: 'deprecated',
+    modification: 'modification',
+    readonly: 'readonly'
+  }
+}
+
+const generateSemanticTokensLegend = (): SemanticTokensLegend => {
+  const tokenTypesLegend = [
```

Comment:
maybe this would be easier to read as
``` typescript
const tokenTypesLegend = Object.keys(TOKEN_LEGEND.types)
tokenTypesLegend.forEach(tokenTypes.set)
```

Reply:
Oh, nice! This shorthand version indeed does the same job. but I think I will keep the original one since it is more explicit. I believe most people are used to seeing an arrow function inside the `forEach` with params clearly passed through and used to perform some tasks.